core-processor/src/ext.rs

```diff
@@ -778,49 +792,56 @@ impl Externalities for Ext {
         pages_num: u32,
         mem: &mut impl Memory,
     ) -> Result<WasmPage, Self::AllocError> {
-        let pages = WasmPage::new(pages_num).map_err(|_| AllocError::ProgramAllocOutOfBounds)?;
-
-        self.context
-            .allocations_context
-            .alloc::<LazyGrowHandler>(pages, mem, |pages| {
-                Ext::charge_gas_if_enough(
-                    &mut self.context.gas_counter,
-                    &mut self.context.gas_allowance_counter,
-                    self.context.page_costs.mem_grow.calc(pages),
-                )
-            })
-            .map_err(Into::into)
+        self.atomic_gas(|ext| {
+            let pages =
+                WasmPage::new(pages_num).map_err(|_| AllocError::ProgramAllocOutOfBounds)?;
+
+            ext.context
+                .allocations_context
+                .alloc::<LazyGrowHandler>(pages, mem, |pages| {
+                    Ext::charge_gas_if_enough(
+                        &mut ext.context.gas_counter,
+                        &mut ext.context.gas_allowance_counter,
+                        ext.context.page_costs.mem_grow.calc(pages),
+                    )
+                })
+                .map_err(Into::into)
+        })
     }
 
     fn free(&mut self, page: WasmPage) -> Result<(), Self::AllocError> {
-        self.context
-            .allocations_context
-            .free(page)
-            .map_err(Into::into)
+        self.atomic_gas(|ext| {
```

Comment:
I guess it's not necessary to use this fn everywhere, even if no gas-related fns called

Reply:
underlying impls can't know anything about gas